### PR TITLE
15420 Update CorrectName unit test and stories

### PR DESCRIPTION
**Pull request by kzdev420**
*Issue #:* /bcgov/entity#15420

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the business-filings-ui license (Apache 2.0).

### CODE REVIEW

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:231]: Is there a good reason that null was last emitted?

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:362]: Re enable

Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts:362]: Also check for others below.

**Review by severinbeauvais — COMMENTED** [review]
Amazing work, thanks!

I'm curious about the commented out lines. Is the code unfinished?

File: tests/unit/CorrectNameRequest.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:231]: Yes.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:349]: Okay

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts:231]: Haha, ok, what is the reason? This event is a boolean but a null is emitted? That doesn't seem right, especially since the line above this verified that the form is valid.

File: tests/unit/CorrectCompanyName.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectCompanyName.spec.ts:82]: I don't understand why the old test verified some data but the new test does not. Did the component functionality change? Why is the company name emitter on line 90 undefined -- shouldn't it have the data?

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:104]: Is the form valid before input? Please update comment on line 103.

Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts:104]: Also on lines 117, 132, 145, 161, 177, 219, etc.

File: tests/unit/CorrectCompanyName.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectCompanyName.spec.ts:82]: I am not sure.
Yesterday it worked with the company name. But, I rebase it with the main br, and it was failed.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:104]: Yes. I will update

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:186]: This should be false, according to the test name (line 173).

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:219]: Why is this commented out?

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:274]: This (and line 271) should match the unit test description (line 234).

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:317]: This (and line 314) should match the unit test description (line 277).

Check the other tests below. I don't want to repeat this comment many times.

The tests need to make sense -- a description of what you're verifying and then proper checks for variables/emitters.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:453]: You need to make this work. This test is a particular scenario and you need to verify all things pertinent to that scenario.

File: tests/unit/CorrectCompanyName.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectCompanyName.spec.ts:21]: Can you use the enum here? (better typing)

File: tests/unit/CorrectNameToNumber.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameToNumber.spec.ts:62]: Check other tests, where we "pop" the last emitted value, since that's the only one we really care about.

I want these tests to actually test that the code is working correctly (from a user point of view). I don't just want the tests to check for whatever happens to be the current value.

File: tests/unit/CorrectNameToNumber.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameToNumber.spec.ts:116]: You need to fix these, otherwise the test is not testing anything.

File: tests/unit/CorrectCompanyName.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectCompanyName.spec.ts:21]: Yes, I can see

File: src/components/correct-name/CorrectCompanyName.vue
Comment (severinbeauvais) [inline src/components/correct-name/CorrectCompanyName.vue:45]: This assignment only happens once, when this component is mounted. Can the companyName prop change after the component is mounted? If so then you should watch the prop instead, eg,
```
@Watch('companyName', { immediate: true })
onCompanyNameChanged (val: string) {
  this.textfield = val
}
```

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:407]: This should be fixed also, otherwise this test doesn't verify much.

File: tests/unit/CorrectCompanyName.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectCompanyName.spec.ts]: Looks ok now.

File: src/components/correct-name/CorrectCompanyName.vue
Comment (kzdev420) [reply src/components/correct-name/CorrectCompanyName.vue:45]: I think it is much better.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:407]: I am on CorrectNameRequest now. I think I can complete shortly

File: tests/unit/CorrectCompanyName.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectCompanyName.spec.ts:23]: Form type is an enum -- you should not initialize it as a string. Use null if needed.

File: tests/unit/CorrectNameToNumber.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectNameToNumber.spec.ts]: Thanks, looks good :)

File: src/components/correct-name/CorrectNameRequest.vue
Comment (kzdev420) [inline src/components/correct-name/CorrectNameRequest.vue:205]: At the test, TypeError: alert is not a function.
The test was not able to run `this.emitSaved(false)`

So, I changed code like this for test case

File: tests/unit/CorrectNameRequest.spec.ts
Comment (kzdev420) [inline tests/unit/CorrectNameRequest.spec.ts:41]: I don't know well how to mock this function.
So, I made the `fetchAndValidateNr` like that based on the test case.

File: src/components/correct-name/CorrectNameRequest.vue
Comment (severinbeauvais) [reply src/components/correct-name/CorrectNameRequest.vue:205]: Could you mock `alert` instead, and check that it has been called?

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts]: ok

**Comment by severinbeauvais** [comment]
Thanks for all the effort you're putting into this! I know it's not always easy.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts]: Please update comment.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:7]: 1. Update comment.
2. Why are you doing this? There may be a valid reason but be aware that this hides console warnings that you might actually want to see when running the unit tests.

Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:7]: Screenshot 2023-07-25 at 2 33 24 PM
To remove this

Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts:7]: Try to fix this by adding this at the top of the unit test file:
```
// Prevent the warning "[Vuetify] Unable to locate target #app"
document.body.setAttribute('id', 'app')
```

Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:7]: I did that. But, when I used that, the following part didn't work
```
await this.showConfirmDialog(
            this.$refs.confirm,
            'Name Request Type Does Not Match Business Type',
            dialogContent
          )
```

Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts:7]: This is getting complicated. Just leave it for now I guess.

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [inline tests/unit/CorrectNameRequest.spec.ts:8]: Please have a comment for each mock here and describe the warning  you're bypassing.

File: tests/unit/CorrectCompanyName.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectCompanyName.spec.ts:41]: This conversation is not yet resolved.

Comment (kzdev420) [reply tests/unit/CorrectCompanyName.spec.ts:46]: I will

File: tests/unit/CorrectNameRequest.spec.ts
Comment (kzdev420) [reply tests/unit/CorrectNameRequest.spec.ts:9]: I will

File: tests/unit/CorrectNameRequest.spec.ts
Comment (severinbeauvais) [reply tests/unit/CorrectNameRequest.spec.ts]: I'm sorry this was so much work but I'm very grateful you got it all done -- thanks!!!

**Review by severinbeauvais — APPROVED** [review]
Well done! This was a pretty deep dive into how the components worked and then how to test them. Kudos!

Approved but please wait for another review or 2.

**Review by JazzarKarim — APPROVED** [review]
LGTM! Great stuff Kevin 👍

**Pull request merged**
